flask_monitoringdashboard/controllers/requests.py

```diff
@@ -30,6 +35,67 @@ def get_num_requests_data(db_session, start_date, end_date):
     }
 
 
+def get_status_code_distribution(db_session, endpoint_id, *criterion):
+    """
+    Gets the distribution of status codes returned by the given endpoint.
+
+    :param db_session: session for the database
+    :param endpoint_id: id for the endpoint
+    :return: A dict where the key is the status code and the value is the fraction of requests that returned the status
+    code. Example: a return value of `{ 200: 0.92, 404: 0.08 }` means that status code 200 was returned on 92% of the
+    requests. 8% of the requests returned a 404 status code.
+    """
+    results = db_session.query(Request.status_code, func.count(Request.status_code)) \
+        .filter(and_(Request.endpoint_id == endpoint_id, Request.status_code.isnot(None), *criterion)) \
+        .group_by(Request.status_code).all()
+
+    total_count = sum(frequency for (_, frequency) in results)
+
+    distribution = {status_code: frequency / total_count for (status_code, frequency) in results}
+
+    return distribution
+
+
+def get_error_requests(db_session, endpoint_id, *criterion):
+    """
+    Gets all requests that did not return a 200 status code.
+
+    :param db_session:
+    :param endpoint_id:
+    :param criterion:
+    :return:
+    """
+    return db_session \
+        .query(Request) \
+        .filter(and_(Request.endpoint_id == endpoint_id, Request.status_code.isnot(None), Request.status_code >= 400,
+                     Request.status_code <= 599, *criterion)) \
+        .all()
+
+
+def get_status_code_frequencies(db_session, endpoint_id, *criterion):
+    """
```

Comment:
the comment is the same as above, but the function is different.

Also, looking at the two functions (distribution and frequency), they are identical except for the past 2 lines. So the query is duplicated. Can you fix this?